codec-dns/src/main/java/io/netty5/handler/codec/dns/DefaultDnsRecordDecoder.java

```diff
@@ -34,75 +34,81 @@ public class DefaultDnsRecordDecoder implements DnsRecordDecoder {
         final long ttl = in.readUnsignedInt();
         final int length = in.readUnsignedShort();
-        final int offset = in.readerIndex();
+        final int offset = in.readerOffset();
 
         if (endOffset - offset < length) {
             // Not enough data
-            in.readerIndex(startOffset);
+            in.readerOffset(startOffset);
             return null;
         }
 
         @SuppressWarnings("unchecked")
         T record = (T) decodeRecord(name, type, aClass, ttl, in, offset, length);
-        in.readerIndex(offset + length);
+        in.readerOffset(offset + length);
         return record;
     }
 
     /**
-     * Decodes a record from the information decoded so far by {@link #decodeRecord(ByteBuf)}.
+     * Decodes a record from the information decoded so far by {@link #decodeRecord(Buffer)}.
      *
      * @param name the domain name of the record
      * @param type the type of the record
      * @param dnsClass the class of the record
      * @param timeToLive the TTL of the record
-     * @param in the {@link ByteBuf} that contains the RDATA
+     * @param in the {@link Buffer} that contains the RDATA
      * @param offset the start offset of the RDATA in {@code in}
      * @param length the length of the RDATA
      *
      * @return a {@link DnsRawRecord}. Override this method to decode RDATA and return other record implementation.
      */
     protected DnsRecord decodeRecord(
             String name, DnsRecordType type, int dnsClass, long timeToLive,
-            ByteBuf in, int offset, int length) throws Exception {
+            Buffer in, int offset, int length) throws Exception {
 
         // DNS message compression means that domain names may contain "pointers" to other positions in the packet
         // to build a full message. This means the indexes are meaningful and we need the ability to reference the
         // indexes un-obstructed, and thus we cannot use a slice here.
         // See https://www.ietf.org/rfc/rfc1035 [4.1.4. Message compression]
-        if (type == DnsRecordType.PTR) {
-            return new DefaultDnsPtrRecord(
-                    name, dnsClass, timeToLive, decodeName0(in.duplicate().setIndex(offset, offset + length)));
+        int roff = in.readerOffset();
+        int woff = in.writerOffset();
+        in.resetOffsets().writerOffset(offset + length).readerOffset(offset);
+        try {
+            if (type == DnsRecordType.PTR) {
+                return new DefaultDnsPtrRecord(
+                        name, dnsClass, timeToLive, decodeName0(in));
+            }
+            if (type == DnsRecordType.CNAME || type == DnsRecordType.NS) {
+                return new DefaultDnsRawRecord(name, type, dnsClass, timeToLive,
+                                               DnsCodecUtil.decompressDomainName(in));
+            }
+            return new DefaultDnsRawRecord(
+                    name, type, dnsClass, timeToLive, in.copy());
```

Comment:
It looked like this method was expected to decode the record without touching the buffer, so I'll go with this defensive approach until the tests otherwise pass, and then we can try it.

Reply:
Using `split()` naively will break parsing of compressed records. I'm getting errors like "name contains a loop". I think getting that to work is beyond the scope of this PR.